h/security/permits/engine.py

```diff
@@ -0,0 +1,69 @@
+from enum import Enum
+
+
+def predicate(requires=None):
+    def decorator(function):
+        function.requires = requires or []
+        return function
+
+    return decorator
+
+
+class PredicateBasedPermissions:
+    def __init__(self, mapping):
+        self.mapping = self._expand_clauses(mapping)
+
+    def permits(self, identity, context, permission):
+        if clauses := self.mapping.get(permission):
+            statement_cache = {}
```

Comment:
It would have to wait until Pyramid 2 but could we use [`RequestLocalCache`](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/request.html?highlight=requestlocalcache#pyramid.request.RequestLocalCache)?